doc/source/DBEP/DBEP02.rst

```diff
@@ -0,0 +1,186 @@
+those documents and leads to several unfortunate situations:
+
+ - *ohpyd* objects hold an instance of an *AssetRegisty*
+ - we need to keep track of **which** *AssertRegistry* things were
+   inserted into
+ - consumers that want access to the asset documents need to also have
+   a handle to the database that the objects are inserting into
+
+The proposed solution is to promote *Resource* and *Datum* documents
+to be peer documents with *Start*, *Stop*, *Descriptor* and *Event*.
+They will appear in the document stream and be inserted into
+*DataBroker* via ``db.insert``.  This eliminates the 'special'
+side-band communication and brings all consumers back to the same
+footing.  This will require coordinated changes to *event-model*,
+*databroker*, *bluesky*, and *ophyd*.
+
+
+
+Implementation
+==============
+
+Currently, *ophyd* is responsible for collecting all of the values for
+the *Resource* and *Datum* documents except for the uids.  The uids
+are generated by calls to ``reg.register_*`` and the datum uids are
+subsequently returned to the *RunEngine* via ``obj.read``.  The
+proposed change is:
+
+ 1. *ophyd* objects would be responsible for generating the full *Resource*
+    and *Datum* documents and providing them to the *RunEngine* to be
+    emitted.  *ophyd* may provide some helpers to make generating compliant
+    documents easy.
+
+    a. Similar to the current documents, a *Resource* must be emitted
+       before any *Datum* that refers to it.  A *Datum* can only refer
+       to a *Resource* that as been emitted after the most recent
+       *Start* and before the *Stop* for the most recent *Start*.
+
+    b. an identical (including uid) *Resource* and *Datum* maybe
+       emitted more than once, the consumers will need to handle this.
+
+    c. The *Datum* documents must be yielded only in the first
+       ``collect_asset_docs`` for which there UID is in ``read``.
+
+    d. The *Resource* documents must only be yielded in the first
+       ``collect_asset_docs`` which includes a *Datum* that refers to
+       it.
+
+
+    e. Calls to ``read`` and ``collect_asset_docs`` must be
+       idempotent.
+
+    Identical *Resource* and *Datum* documents are to support a single
+    *Resource* that may span many runs, such as background images, and
+    still ensure that with in the scope of a *Start* / *Stop* pair a
+    consumer will see all of the documents required.
+
+
+
+ 2. in ``save`` before the *Event* document is emitted the *RunEngine*
+    will acquire and emit any *AssetRegistry* Documents.
+
+    a. in ``save`` the *RunEngine* knows what objects in the bundle, call
+       ``collect_asset_docs`` method ::
+
+            def collect_asset_docs(self) -> Iterator[Tuple[str, Dict[str, Any]]]:
+                ...
+
+       which will yield the ``(name, doc)`` pairs for anything that
+       was just read.
+
+    b. these documents will be emitted **before** the *Event*
```

Comment:
Does this mean that the raw stream will be:
1. Start
1. Descriptor
1. Resource
1. Datum
1. Event
1. Datum
1. Event
1. Stop

Does that mean that there will be two sources of the "data", one in the Datums and one in the Events?

Reply:
Yes, that is the order.



> Does that mean that there will be two sources of the "data", one in the Datums and one in the Events?

"data" is too ambiguous in this sentence :)  

Currently, there is the values in the event and then a magical asset registry (nee filestore) which lets you trade tokens for (possibly large) arrays stored outside of the database.   This is just surfacing the information that is in the asset registry in a way that all clients can see (and plugs the _huge_ hole in our conceptual model that having ophyd objects know about an asset registry).  This should make it strait forward to write a consumer that ex-filtrates everything (documents and files) to an external system (for example Simon's AWS account :smiling_imp: ). 

From the point of view of 'source' in the descriptor there will still only be one value (as it is now).

Reply:
Sorry in this example "data" are the images coming from an area detector.
Maybe a better question is will I still be able to/have to trade tokens for arrays?
For example does `LiveImage` still look for events or datums?
Do all callbacks which want large arrays need to cache datums and then wait for their event to come down?

Reply:
`LiveImage` now has two options.

1. It can continue to work the way it does today, ignoring the new documents (datum and resource) that flow through and trading tokens for arrays via a `Broker`. A downside of this approach is that sometimes the RunEngine hands `LiveImage` an Event with some datum_id _before_ the `Registry` knows about it. (MongoDB takes some time to reflect new information in queries.) To be robust, `LiveImage` has to be prepared to poll `Broker` about a `datum_id` more than once, waiting for things to sync.
2. Alternatively, `LiveImage` can cache the datum and resource documents and use a Handler directly to load the data. In this scenario, `LiveImage` doesn't need a `Broker` (there doesn't even technically need to _be_ one!), but it does now need a Handler.

Reply:
This discussion points on the handler registry, management, and caching should probably be re-factored to make option 2 easier.